src/summaries.rs

```diff
@@ -98,8 +98,26 @@ pub fn summarize(
 /// the summary cache, which is a key value store. The string will always be the same as
 /// long as the definition does not change its name or location, so it can be used to
 /// transfer information from one compilation to the next, making incremental analysis possible.
-pub fn summary_key_str(tcx: &TyCtxt, crate_name: &str, def_id: DefId) -> String {
-    let mut name: String = String::from(crate_name);
+pub fn summary_key_str(tcx: &TyCtxt, def_id: DefId) -> String {
+    let crate_name = if def_id.is_local() {
+        tcx.crate_name.as_interned_str().as_str().get()
+    } else {
+        // This is both ugly and probably brittle, but I can't find any other
+        // way to retrieve the crate name from a def_id that is not local.
+        // tcx.crate_data_as_rc_any returns an untracked value, which is potentially problematic
+        // as per the comments on the function:
+        // "Note that this is *untracked* and should only be used within the query
+        // system if the result is otherwise tracked through queries"
+        // For now, this seems OK since we are only using the crate name.
+        // Of course, should a crate name change in an incremental scenario this
+        // is going to be the least of our worries.
+        let cdata = tcx.crate_data_as_rc_any(def_id.krate);
+        let cdata = cdata
+            .downcast_ref::<rustc_metadata::cstore::CrateMetadata>()
+            .unwrap();
+        cdata.name.as_str().get()
+    };
+    let mut name = String::from(crate_name);
```

Comment:
I am a bit conflicted about this. String::from just calls crate_name.to_owned(), so why have the middle man in the picture? Mostly because it makes it obvious that a new String is being created. On the other hand, it is less obvious that crate_name is being consumed. I'm still learning best practice here, so I don't have a strong opinion. But using String::from is still my current first instinct.

Reply:
Yeah I don't think there is any universally agreed on version. There is some discussion here: https://users.rust-lang.org/t/what-is-the-idiomatic-way-to-convert-str-to-string/12160/10. I'm somewhat convinced of the to_owned() argument, but it doesn't matter much.